pkg/cloud/services/ec2/instances.go

```diff
@@ -1307,3 +1324,37 @@ func getInstanceCPUOptionsRequest(cpuOptions infrav1.CPUOptions) *types.CpuOptio
 
 	return request
 }
+
+// shouldEnablePrimaryIpv6 determines whether to enable a primary IPv6 address for an instance.
+// It returns true if both the VPC has IPv6 enabled and the instance's subnet has an IPv6 CIDR block.
+// This is required when registering instances by ID to IPv6 target groups.
+func (s *Service) shouldEnablePrimaryIpv6(i *infrav1.Instance) (bool, error) {
+	var enablePrimaryIpv6 bool
+
+	// We should enable IPv6 capabilities only when the users explicitly configure so.
+	if !s.scope.VPC().IsIPv6Enabled() {
+		return false, nil
+	}
+
+	sn := s.scope.Subnets().FindByID(i.SubnetID)
+	if sn != nil {
+		enablePrimaryIpv6 = sn.IsIPv6
+	} else {
+		// The subnet is in a different VPC than the cluster VPC. Then, we query AWS API.
+		sns, err := s.getFilteredSubnets(types.Filter{Name: aws.String("subnet-id"), Values: []string{i.SubnetID}})
+		if err != nil {
+			return false, fmt.Errorf("failed to find subnet info with id %q for instance: %w", i.SubnetID, err)
+		}
+		if len(sns) == 0 {
+			return false, fmt.Errorf("expected subnet %q for instance to exist, but found none", i.SubnetID)
+		}
+		for _, set := range sns[0].Ipv6CidrBlockAssociationSet {
```

Comment:
Since we're filtering by ID, `len(sns) == 1` should always be true, but I wonder if we want to have a failsafe check on that.

Reply:
Right, addressed!